adapters/coap-vertx-base/src/main/java/org/eclipse/hono/adapter/coap/DeviceInfoSupplier.java

```diff
@@ -0,0 +1,135 @@
+import java.util.Objects;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import org.eclipse.californium.elements.auth.AdditionalInfo;
+import org.eclipse.californium.scandium.auth.AdvancedApplicationLevelInfoSupplier;
+import org.eclipse.californium.scandium.auth.ApplicationLevelInfoSupplier;
+import org.eclipse.hono.auth.Device;
+import org.eclipse.hono.tracing.TracingHelper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.opentracing.Span;
+import io.opentracing.Tracer;
+import io.opentracing.tag.Tags;
+
+/**
+ * Device information supplier.
+ */
+public class DeviceInfoSupplier implements ApplicationLevelInfoSupplier, AdvancedApplicationLevelInfoSupplier {
+
+    /**
+     * Key in the extended info of the peer identity containing the auth-id used for authentication.
+     */
+    public static final String EXT_INFO_KEY_HONO_AUTH_ID = "hono-auth-id";
+    /**
+     * Key in the extended info of the peer identity containing the authenticated Device.
+     */
+    public static final String EXT_INFO_KEY_HONO_DEVICE = "hono-device";
+
+    private static final Logger LOG = LoggerFactory.getLogger(DeviceInfoSupplier.class);
+
+    /**
+     * The vert.x context to run interactions with Hono services on.
+     */
+    private final Tracer tracer;
+    private final String adapterName;
+    private final List<AdvancedApplicationLevelInfoSupplier> suppliers = new CopyOnWriteArrayList<>();
+
+    /**
+     * Creates a new resolver.
+     *
+     * @param tracer The OpenTracing tracer.
+     * @param adapterName The name of the protocol adapter.
+     * @throws NullPointerException if any of the parameters are {@code null}.
+     */
+    public DeviceInfoSupplier(
+            final Tracer tracer,
+            final String adapterName) {
+
+        this.tracer = Objects.requireNonNull(tracer);
+        this.adapterName = Objects.requireNonNull(adapterName);
+    }
+
+    /**
+     * Add authentication specific supplier.
+     *
+     * Used on missing {@code customArgument} in {@link #getInfo(Principal, Object)}.
+     *
+     * @param supplier authentication specific supplier
+     */
+    public void add(final AdvancedApplicationLevelInfoSupplier supplier) {
+        suppliers.add(supplier);
+    }
+
+    private Span newSpan(final String operation) {
+        return tracer.buildSpan(operation)
+                .withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_SERVER)
+                .withTag(Tags.COMPONENT.getKey(), adapterName)
+                .start();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public AdditionalInfo getInfo(final Principal clientIdentity) {
+        return getInfo(clientIdentity, null);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public AdditionalInfo getInfo(final Principal clientIdentity, final Object customArgument) {
+        if (customArgument instanceof AdditionalInfo) {
+            final Span newSpan = newSpan("get device info");
```

Comment:
this span is not getting finished, is it? It doesn't really seem helpful to create a Span here in the first place as there is no real processing to be traced, is there? It should be sufficient to simply let a client log the outcome of this method in a span, if necessary ....

Reply:
See my comment above ... if there is a decision about the resumption and behavior, that may make it easier.

Reply:
[This one](https://github.com/eclipse/hono/pull/2629#discussion_r621280534).

You may also consider to read [questions](https://github.com/eclipse/hono/pull/2629#issuecomment-827611731).

> What purpose does the span have?

That depends on the intended functionality based on the (short/mid term) requirements. Therefore the questions.
  
> Why do you want to log an error if the passed in object is an AdditionalInfo instance?

The cleanup (instead of delete) depends on the answers.